File: sdks/go/cmd/elydora/plugins/droid_transaction_test.go
package plugins

import (
	"errors"
	"os"
	"strings"
	"testing"
)

func prepareDroidTestInstallation(
	t *testing.T,
	fixture *droidFixture,
	config InstallConfig,
) *preparedDroidInstallation {
	t.Helper()
	sources, err := readDroidSources()
	if err != nil {
		t.Fatalf("read Factory Droid sources: %v", err)
	}
	paths, nodePath, err := preflightDroidInstallation(config, sources)
	if err != nil {
		t.Fatalf("preflight Factory Droid installation: %v", err)
	}
	rendered, err := renderDroidInstallation(
		sources,
		paths.guardPath,
		paths.auditPath,
		nodePath,
		paths.runtimeRoot,
	)
	if err != nil {
		t.Fatalf("render Factory Droid installation: %v", err)
	}
	prepared, err := prepareDroidInstallation(config, sources, rendered)
	if err != nil {
		t.Fatalf("prepare Factory Droid installation: %v", err)
	}
	return prepared
}

func droidManagedPaths(fixture *droidFixture) []string {
	return []string{
		fixture.guardPath,
		fixture.runtimeConfig,
		fixture.privateKey,
		fixture.hookPath,
		fixture.configPath,
	}
}

func TestDroidInstallRollsBackRuntimeAfterLateHookCommitFailure(t *testing.T) {
	fixture := prepareDroidFixture(t, droidFixtureOptions{})
	prepared := prepareDroidTestInstallation(t, fixture, fixture.config)
	failed := false
	rename := func(source, destination string) error {
		if !failed && sameDroidPath(destination, fixture.configPath) {
			failed = true
			return errors.New("injected Droid hook commit failure")
		}
		return os.Rename(source, destination)
	}
	err := commitDroidInstallation(prepared, rename)
	if err == nil || !strings.Contains(err.Error(), "injected Droid hook commit failure") || !failed {
		t.Fatalf("transaction error = %v, failed = %v", err, failed)
	}
	for _, path := range droidManagedPaths(fixture) {
		requireMissingDroidFile(t, path)
	}
	requireNoDroidStagingFiles(t, fixture.homeDir)
}

func TestDroidConcurrentActiveHookReplacementPreventsRuntimeCommit(t *testing.T) {
	fixture := prepareDroidFixture(t, droidFixtureOptions{})
	installDroidFixture(t, fixture)
	before := snapshotDroidFiles(t, droidManagedPaths(fixture)...)
	updated := fixture.config
	updated.OrgID = "org-updated"
	prepared := prepareDroidTestInstallation(t, fixture, updated)
	concurrent := "{\"hooks\":{\"PreToolUse\":[]},\"owner\":\"concurrent\"}\n"
	if err := os.WriteFile(fixture.configPath, []byte(concurrent), 0600); err != nil {
		t.Fatalf("replace active hook source: %v", err)
	}
	err := commitDroidInstallation(prepared, nil)
	if err == nil || !strings.Contains(err.Error(), "changed during Install Factory Droid") {
		t.Fatalf("concurrent source error = %v", err)
	}
	for path, source := range before {
		if sameDroidPath(path, fixture.configPath) {
			continue
		}
		if readDroidTestFile(t, path) != source {
			t.Fatalf("runtime changed after concurrent source replacement: %s", path)
		}
	}
	if readDroidTestFile(t, fixture.configPath) != concurrent {
		t.Fatal("concurrent source was overwritten")
	}
	requireNoDroidStagingFiles(t, fixture.homeDir)
}

func TestDroidConcurrentInactiveLocalSourceCreationIsDetected(t *testing.T) {
	fixture := prepareDroidFixture(t, droidFixtureOptions{})
	installDroidFixture(t, fixture)
	before := snapshotDroidFiles(t, droidManagedPaths(fixture)...)
	updated := fixture.config
	updated.OrgID = "org-updated"
	prepared := prepareDroidTestInstallation(t, fixture, updated)
	concurrent := "{\"hooksDisabled\":true}\n"
	writeOptionalDroidFile(t, fixture.localSettingsPath, droidString(concurrent))
	err := commitDroidInstallation(prepared, nil)
	if err == nil || !strings.Contains(err.Error(), "local settings changed during Install") {
		t.Fatalf("concurrent local source error = %v", err)
	}
	requireDroidSnapshot(t, before)
	if readDroidTestFile(t, fixture.localSettingsPath) != concurrent {
		t.Fatal("concurrent local settings source was overwritten")
	}
	requireNoDroidStagingFiles(t, fixture.homeDir)
}

func TestDroidConcurrentProjectPolicyChangeIsDetected(t *testing.T) {
	fixture := prepareDroidFixture(t, droidFixtureOptions{
		projectSettings: droidJSON(map[string]any{"hooksDisabled": false}),
	})
	installDroidFixture(t, fixture)
	before := snapshotDroidFiles(t, droidManagedPaths(fixture)...)
	updated := fixture.config
	updated.OrgID = "org-updated"
	prepared := prepareDroidTestInstallation(t, fixture, updated)
	concurrent := "{\"hooksDisabled\":true}\n"
	if err := os.WriteFile(fixture.projectSettingsPath, []byte(concurrent), 0600); err != nil {
		t.Fatalf("change project policy: %v", err)
	}
	err := commitDroidInstallation(prepared, nil)
	if err == nil || !strings.Contains(err.Error(), "project settings changed during Install") {
		t.Fatalf("concurrent project policy error = %v", err)
	}
	requireDroidSnapshot(t, before)
	if readDroidTestFile(t, fixture.projectSettingsPath) != concurrent {
		t.Fatal("concurrent project policy was overwritten")
	}
	requireNoDroidStagingFiles(t, fixture.homeDir)
}

func TestDroidConcurrentPolicyChangePreventsRuntimeDirectoryCreation(t *testing.T) {
	fixture := prepareDroidFixture(t, droidFixtureOptions{
		projectSettings: droidJSON(map[string]any{"hooksDisabled": false}),
	})
	prepared := prepareDroidTestInstallation(t, fixture, fixture.config)
	if err := os.WriteFile(
		fixture.projectSettingsPath,
		[]byte("{\"hooksDisabled\":true}\n"),
		0600,
	); err != nil {
		t.Fatalf("change project policy: %v", err)
	}
	err := commitDroidInstallation(prepared, nil)
	if err == nil || !strings.Contains(err.Error(), "project settings changed during Install") {
		t.Fatalf("concurrent project policy error = %v", err)
	}
	requireMissingDroidFile(t, fixture.agentDir)
	requireMissingDroidFile(t, fixture.configPath)
}

func TestDroidSameContentReplacementIsRejectedByPhysicalIdentity(t *testing.T) {
	fixture := prepareDroidFixture(t, droidFixtureOptions{})
	installDroidFixture(t, fixture)
	before := snapshotDroidFiles(t, droidManagedPaths(fixture)...)
	sources, err := readDroidSources()
	if err != nil {
		t.Fatalf("read Factory Droid sources: %v", err)
	}
	original := readDroidTestFile(t, fixture.configPath)
	if err := os.Remove(fixture.configPath); err != nil {
		t.Fatalf("remove hook source: %v", err)
	}
	if err := os.WriteFile(fixture.configPath, []byte(original), 0600); err != nil {
		t.Fatalf("replace hook source: %v", err)
	}
	updated := fixture.config
	updated.OrgID = "org-updated"
	paths, nodePath, err := preflightDroidInstallation(updated, sources)
	if err != nil {
		t.Fatalf("preflight stale sources: %v", err)
	}
	rendered, err := renderDroidInstallation(
		sources,
		paths.guardPath,
		paths.auditPath,
		nodePath,
		paths.runtimeRoot,
	)
	if err != nil {
		t.Fatalf("render stale sources: %v", err)
	}
	prepared, err := prepareDroidInstallation(updated, sources, rendered)
	if err != nil {
		t.Fatalf("prepare stale sources: %v", err)
	}
	err = commitDroidInstallation(prepared, nil)
	if err == nil || !strings.Contains(err.Error(), "changed during Install") {
		t.Fatalf("same-content replacement error = %v", err)
	}
	if readDroidTestFile(t, fixture.configPath) != original {
		t.Fatal("same-content replacement changed")
	}
	for path, source := range before {
		if !sameDroidPath(path, fixture.configPath) && readDroidTestFile(t, path) != source {
			t.Fatalf("runtime changed after inode replacement: %s", path)
		}
	}
	requireNoDroidStagingFiles(t, fixture.homeDir)
}

func TestDroidUninstallRollsBackAllHookDocuments(t *testing.T) {
	fixture := prepareDroidFixture(t, droidFixtureOptions{})
	installDroidFixture(t, fixture)
	installed := droidCurrentHooks(t, fixture.configPath)
	writeDroidTestObject(t, fixture.settingsPath, map[string]any{
		"hooks": installed,
		"owner": "user",
	})
	paths := append(droidManagedPaths(fixture), fixture.settingsPath)
	before := snapshotDroidFiles(t, paths...)
	sources, err := readDroidSources()
	if err != nil {
		t.Fatalf("read uninstall sources: %v", err)
	}
	runtimeRoot, err := droidRuntimeRoot()
	if err != nil {
		t.Fatalf("resolve runtime root: %v", err)
	}
	rendered, err := renderDroidUninstall(sources, droidTestAgentID, runtimeRoot)
	if err != nil {
		t.Fatalf("render uninstall: %v", err)
	}
	prepared, err := prepareDroidUninstall(rendered)
	if err != nil {
		t.Fatalf("prepare uninstall: %v", err)
	}
	failed := false
	rename := func(source, destination string) error {
		if !failed && sameDroidPath(destination, fixture.settingsPath) {
			failed = true
			return errors.New("injected Droid uninstall failure")
		}
		return os.Rename(source, destination)
	}
	err = commitDroidUninstall(prepared, rename)
	if err == nil || !strings.Contains(err.Error(), "injected Droid uninstall failure") || !failed {
		t.Fatalf("uninstall rollback error = %v, failed = %v", err, failed)
	}
	requireDroidSnapshot(t, before)
	requireNoDroidStagingFiles(t, fixture.homeDir)
}

func TestDroidPreflightBlocksDisabledHooksBeforeRuntimeCreation(t *testing.T) {
	fixture := prepareDroidFixture(t, droidFixtureOptions{
		settings: droidJSON(map[string]any{"hooksDisabled": true}),
	})
	preflighter, ok := any(fixture.plugin).(InstallPreflighter)
	if !ok {
		t.Fatal("Factory Droid plugin does not implement InstallPreflighter")
	}
	err := preflighter.PreflightInstall(fixture.config)
	if err == nil || !strings.Contains(err.Error(), "hooksDisabled") {
		t.Fatalf("preflight error = %v", err)
	}
	requireMissingDroidFile(t, fixture.agentDir)
}

File: sdks/node/test/droid-transaction.test.mjs
import assert from 'node:assert/strict';
import { mkdir, readFile, rm, writeFile } from 'node:fs/promises';
import path from 'node:path';
import test from 'node:test';
import {
  assertMissing,
  assertNoTransactionFiles,
  configModuleUrl,
  contractModuleUrl,
  createFixture,
  environment,
  installConfig,
  installationModuleUrl,
  ioModuleUrl,
  runNode,
  writeConfig,
} from '../test-support/droid-test-helpers.mjs';

const managedFiles = ['guard.js', 'config.json', 'private.key', 'hook.js'];

async function snapshotInstallation(fixture, extraPaths = []) {
  const paths = [
    ...managedFiles.map((name) => path.join(fixture.agentDir, name)),
    fixture.rootPath,
    ...extraPaths,
  ];
  return new Map(await Promise.all(paths.map(async (filePath) => [
    filePath,
    await readFile(filePath, 'utf-8'),
  ])));
}

async function assertSnapshot(snapshot) {
  for (const [filePath, source] of snapshot) {
    assert.equal(await readFile(filePath, 'utf-8'), source, filePath);
  }
}

function transactionEnvironment(fixture, overrides = {}) {
  return {
    ...environment(fixture),
    ELYDORA_INSTALL_CONFIG: JSON.stringify(installConfig(fixture, overrides)),
  };
}

const readAndRenderSource = `
  import { buildGroup } from ${JSON.stringify(contractModuleUrl)};
  import {
    activeDocument,
    additionsForTarget,
    installationDocuments,
    renderDocument,
  } from ${JSON.stringify(configModuleUrl)};
  import { readSources } from ${JSON.stringify(ioModuleUrl)};
  import {
    preflightDroidInstallation,
    prepareDroidInstallation,
  } from ${JSON.stringify(installationModuleUrl)};
  const config = JSON.parse(process.env.ELYDORA_INSTALL_CONFIG);
  const sources = await readSources();
`;

const prepareSource = `${readAndRenderSource}
  const paths = await preflightDroidInstallation(config, sources);
  const target = activeDocument(sources);
  const groups = new Map([
    ['PreToolUse', buildGroup(paths.guardPath)],
    ['PostToolUse', buildGroup(paths.auditPath)],
  ]);
  const rendered = installationDocuments(sources).map((document) => renderDocument(
    document,
    undefined,
    additionsForTarget(document, target, groups),
  ));
  const prepared = await prepareDroidInstallation(config, sources, rendered);
`;

test('Droid rolls back runtime and hooks after a late commit failure', async () => {
  const fixture = await createFixture();
  try {
    assert.equal((await fixture.install()).code, 0);
    const before = await snapshotInstallation(fixture);
    const source = `${prepareSource}
      import { rename } from 'node:fs/promises';
      import { commitDroidInstallation } from ${JSON.stringify(installationModuleUrl)};
      let calls = 0;
      await commitDroidInstallation(prepared, async (from, to) => {
        calls += 1;
        if (calls === 5) throw new Error('injected Droid rename failure');
        await rename(from, to);
      });
    `;
    const result = await runNode(
      ['--input-type=module', '--eval', source],
      transactionEnvironment(fixture, { orgId: 'org-updated', token: 'token-updated' }),
      fixture.workspaceDir,
    );
    assert.equal(result.code, 1);
    assert.match(result.stderr, /injected Droid rename failure/i);
    await assertSnapshot(before);
    await assertNoTransactionFiles(fixture);
  } finally {
    await fixture.close();
  }
});

test('Droid detects concurrent active hook replacement before runtime commit', async () => {
  const fixture = await createFixture();
  try {
    assert.equal((await fixture.install()).code, 0);
    const before = await snapshotInstallation(fixture);
    const concurrent = '{"hooks":{"PreToolUse":[]},"owner":"concurrent"}\n';
    const source = `${prepareSource}
      import { writeFile } from 'node:fs/promises';
      import { commitDroidInstallation } from ${JSON.stringify(installationModuleUrl)};
      await writeFile(process.env.ELYDORA_CONCURRENT_PATH, process.env.ELYDORA_CONCURRENT);
      await commitDroidInstallation(prepared);
    `;
    const result = await runNode(
      ['--input-type=module', '--eval', source],
      {
        ...transactionEnvironment(fixture, { orgId: 'org-updated' }),
        ELYDORA_CONCURRENT_PATH: fixture.rootPath,
        ELYDORA_CONCURRENT: concurrent,
      },
      fixture.workspaceDir,
    );
    assert.equal(result.code, 1);
    assert.match(result.stderr, /hooks changed during Factory Droid installation/i);
    for (const [filePath, original] of before) {
      if (filePath !== fixture.rootPath) assert.equal(await readFile(filePath, 'utf-8'), original);
    }
    assert.equal(await readFile(fixture.rootPath, 'utf-8'), concurrent);
    await assertNoTransactionFiles(fixture);
  } finally {
    await fixture.close();
  }
});

test('Droid tracks concurrent creation of an inactive local settings source', async () => {
  const fixture = await createFixture();
  try {
    assert.equal((await fixture.install()).code, 0);
    const before = await snapshotInstallation(fixture);
    const concurrent = '{"hooksDisabled":true}\n';
    const source = `${prepareSource}
      import { writeFile } from 'node:fs/promises';
      import { commitDroidInstallation } from ${JSON.stringify(installationModuleUrl)};
      await writeFile(process.env.ELYDORA_LOCAL_PATH, process.env.ELYDORA_CONCURRENT);
      await commitDroidInstallation(prepared);
    `;
    const result = await runNode(
      ['--input-type=module', '--eval', source],
      {
        ...transactionEnvironment(fixture, { orgId: 'org-updated' }),
        ELYDORA_LOCAL_PATH: fixture.localSettingsPath,
        ELYDORA_CONCURRENT: concurrent,
      },
      fixture.workspaceDir,
    );
    assert.equal(result.code, 1);
    assert.match(result.stderr, /local settings changed during Factory Droid installation/i);
    await assertSnapshot(before);
    assert.equal(await readFile(fixture.localSettingsPath, 'utf-8'), concurrent);
    await assertNoTransactionFiles(fixture);
  } finally {
    await fixture.close();
  }
});

test('Droid tracks concurrent changes to read-only project policy', async () => {
  const fixture = await createFixture({ projectSettings: { hooksDisabled: false } });
  try {
    assert.equal((await fixture.install()).code, 0);
    const before = await snapshotInstallation(fixture);
    const concurrent = '{"hooksDisabled":true}\n';
    const source = `${prepareSource}
      import { writeFile } from 'node:fs/promises';
      import { commitDroidInstallation } from ${JSON.stringify(installationModuleUrl)};
      await writeFile(process.env.ELYDORA_POLICY_PATH, process.env.ELYDORA_CONCURRENT);
      await commitDroidInstallation(prepared);
    `;
    const result = await runNode(
      ['--input-type=module', '--eval', source],
      {
        ...transactionEnvironment(fixture, { orgId: 'org-updated' }),
        ELYDORA_POLICY_PATH: fixture.projectSettingsPath,
        ELYDORA_CONCURRENT: concurrent,
      },
      fixture.workspaceDir,
    );
    assert.equal(result.code, 1);
    assert.match(result.stderr, /project settings changed during Factory Droid installation/i);
    await assertSnapshot(before);
    assert.equal(await readFile(fixture.projectSettingsPath, 'utf-8'), concurrent);
    await assertNoTransactionFiles(fixture);
  } finally {
    await fixture.close();
  }
});

test('Droid rejects same-content stale hook snapshots by physical identity', async () => {
  const fixture = await createFixture();
  try {
    assert.equal((await fixture.install()).code, 0);
    const before = await snapshotInstallation(fixture);
    const source = `${readAndRenderSource}
      import { readFile, rm, writeFile } from 'node:fs/promises';
      const original = await readFile(sources.root.filePath, 'utf-8');
      await rm(sources.root.filePath);
      await writeFile(sources.root.filePath, original);
      const paths = await preflightDroidInstallation(config, sources);
      const target = activeDocument(sources);
      const groups = new Map([
        ['PreToolUse', buildGroup(paths.guardPath)],
        ['PostToolUse', buildGroup(paths.auditPath)],
      ]);
      const rendered = installationDocuments(sources).map((document) => renderDocument(
        document,
        undefined,
        additionsForTarget(document, target, groups),
      ));
      await prepareDroidInstallation(config, sources, rendered);
    `;
    const result = await runNode(
      ['--input-type=module', '--eval', source],
      transactionEnvironment(fixture),
      fixture.workspaceDir,
    );
    assert.equal(result.code, 1);
    assert.match(result.stderr, /hooks changed before installation/i);
    await assertSnapshot(before);
    await assertNoTransactionFiles(fixture);
  } finally {
    await fixture.close();
  }
});

test('Droid preserves orphaned runtime artifacts without a verifiable identity', async () => {
  const fixture = await createFixture();
  try {
    await mkdir(fixture.agentDir, { recursive: true });
    await writeFile(fixture.guardScriptPath, 'orphaned guard\n');
    const result = await fixture.install();
    assert.equal(result.code, 1);
    assert.match(result.stderr, /identity cannot be verified/i);
    assert.equal(await readFile(fixture.guardScriptPath, 'utf-8'), 'orphaned guard\n');
    await assertMissing(fixture.rootPath);
  } finally {
    await fixture.close();
  }
});

test('Droid uninstall restores all hook documents after a later removal fails', async () => {
  const fixture = await createFixture();
  try {
    assert.equal((await fixture.install()).code, 0);
    const installed = JSON.parse(
      (await readFile(fixture.rootPath, 'utf-8')).replace(/^\/\/[^\n]*\n/, ''),
    );
    await writeConfig(fixture.settingsPath, { hooks: installed.hooks, owner: 'user' });
    const before = await snapshotInstallation(fixture, [fixture.settingsPath]);
    const source = `
      import { renderDocument, sourceDocuments } from ${JSON.stringify(configModuleUrl)};
      import { readSources } from ${JSON.stringify(ioModuleUrl)};
      import {
        commitDroidUninstall,
        prepareDroidUninstall,
      } from ${JSON.stringify(installationModuleUrl)};
      import { rename } from 'node:fs/promises';
      const sources = await readSources();
      const rendered = sourceDocuments(sources).map((document) => renderDocument(
        document,
        'agent-1',
        new Map(),
      ));
      const prepared = await prepareDroidUninstall(rendered);
      let calls = 0;
      await commitDroidUninstall(prepared, async (from, to) => {
        calls += 1;
        if (calls === 2) throw new Error('injected Droid uninstall failure');
        await rename(from, to);
      });
    `;
    const result = await runNode(
      ['--input-type=module', '--eval', source],
      environment(fixture),
      fixture.workspaceDir,
    );
    assert.equal(result.code, 1);
    assert.match(result.stderr, /injected Droid uninstall failure/i);
    await assertSnapshot(before);
    await assertNoTransactionFiles(fixture);
  } finally {
    await fixture.close();
  }
});
